home/utils/useUserReviewCheck.ts

```diff
@@ -0,0 +1,102 @@
+
+export function listenForForegroundEvent(
+  listener: (event: any) => Promise<any>
+): EventSubscription {
+  return addListenerWithNativeCallback('ExponentKernel.foregroundHome', listener);
+}
+
+type UserReviewInfo = {
+  hasAskedForNativeReview: boolean;
+  lastDismissDate?: Date;
+  appOpenedCounter: number;
+};
+
+const userReviewInfoDefaultValues: UserReviewInfo = {
+  hasAskedForNativeReview: false,
+  lastDismissDate: undefined,
+  appOpenedCounter: 0,
+};
+
+type UseUserReviewCheckParams = {
+  apps?: CommonAppDataFragment[];
+  snacks?: CommonSnackDataFragment[];
+};
+
+export const useUserReviewCheck = ({ apps, snacks }: UseUserReviewCheckParams) => {
+  const [lastCrashDate, setLastCrashDate] = useState<Date>();
+  const [userReviewInfo, setUserReviewInfo] = useState<UserReviewInfo>();
+
+  const timeNow = new Date();
+  const noRecentCrashes = lastCrashDate
+    ? timeNow.getTime() - new Date(lastCrashDate).getTime() > 60 * 60 * 1000
+    : true;
+  const noRecentDismisses = userReviewInfo?.lastDismissDate
+    ? timeNow.getTime() - userReviewInfo?.lastDismissDate.getTime() > 15 * 24 * 60 * 60 * 1000
+    : true;
+
+  /**
+   * We should only prompt users to review the app if they seem to be
+   * having a good experience, to check that we verify if the user
+   * has not experienced any crashes in the last hour, and has at least
+   * 5 apps or 5 snacks or has opened the app at least 50 times.
+   * If the user dismisses the review section, we only show it again
+   * after 15 days.
+   */
+  const shouldShowReviewSection =
+    userReviewInfo &&
+    !userReviewInfo.hasAskedForNativeReview &&
+    noRecentCrashes &&
+    noRecentDismisses &&
+    (userReviewInfo.appOpenedCounter >= 50 ||
+      (apps?.length || 0) >= 5 ||
+      (snacks?.length || 0) >= 5);
+
+  function updateUserReviewInfo(info: Partial<UserReviewInfo>) {
+    const newInfo: UserReviewInfo = { ...userReviewInfoDefaultValues, ...userReviewInfo, ...info };
+    AsyncStorage.setItem('userReviewInfo', JSON.stringify(userReviewInfo));
+    setUserReviewInfo(newInfo);
+  }
+
+  useEffect(() => {
+    AsyncStorage.getItem('userReviewInfo').then((info) => {
+      const userReviewInfo: UserReviewInfo = info ? JSON.parse(info) : {};
+      userReviewInfo.appOpenedCounter = Number(userReviewInfo.appOpenedCounter || 0) + 1;
+      updateUserReviewInfo(userReviewInfo);
+    });
+  }, []);
+
+  useEffect(() => {
+    const getLastCrashDate = async () => {
+      const lastCrash = await Kernel.getLastCrashDate();
+      setLastCrashDate(lastCrash ? new Date(lastCrash) : undefined);
+    };
+    const listener = listenForForegroundEvent(getLastCrashDate);
+
+    getLastCrashDate();
+
+    return listener.remove;
+  }, []);
+
+  function requestStoreReview() {
+    updateUserReviewInfo({ hasAskedForNativeReview: true });
+    StoreReview.requestReview();
+  }
+
+  function dismissReviewSection() {
+    updateUserReviewInfo({ lastDismissDate: new Date() });
```

Comment:
It would be beneficial to provide some way for users to provide constructive criticism here. It could e.g. be an in-app form that we send to GA, or a link to a google form/typeform, an email.

Reply:
Yeah, I like the in-app form idea, let me give this a try

Reply:
Here is a very basic feedback design that we could use.

<img width="259" alt="image" src="https://github.com/expo/expo/assets/11707729/cf9aa49a-6306-4e85-93c2-8fb777643370">

In terms of where to send these feedbacks to I'll be opening a PR in /universe to add a feedback route to expo-go, kinda the same way we did here https://github.com/expo/universe/pull/14010

Reply:
Nice, I think that would work perfectly! 👌